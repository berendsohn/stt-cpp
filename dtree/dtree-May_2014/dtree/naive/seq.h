// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_SEQ_H_
#define DTREE_NAIVE_SEQ_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndSeq : public Base {
 public:
  EndSeq() : solid_() {}
  EndSeq* solid(int i) const { return solid_[i]; }
  void set_solid(int i, EndSeq* solid_i) { solid_[i] = solid_i; }
  template<typename Forwarder>
  void CopyFrom(const EndSeq& node, Forwarder forward);
 private:
  EndSeq* solid_[2];
  DISALLOW_COPY_AND_ASSIGN(EndSeq);
};

template<typename Group, typename Base>
class WithValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToSeq(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromSeq(Node* node, typename Group::Type delta);

 protected:
  WithValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  template<typename Node>
  static typename Semigroup::Type AggrSeq(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostSeq(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeftmostSeq(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRightmostSeq(Node* node, const Predicate& predicate);
 protected:
  WithAggr() {}
};

template<typename Base>
class WithReverseBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverseBy() {}
};

template<typename Base>
class WithReverse : public Base {
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverse() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_SEQ_H_
