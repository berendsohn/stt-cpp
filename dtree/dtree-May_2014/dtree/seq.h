// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_SEQ_H_
#define DTREE_SEQ_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/seq.h"
#else
#include "dtree/selfadjust/seq.h"
#endif
#endif  // DTREE_SEQ_H_
