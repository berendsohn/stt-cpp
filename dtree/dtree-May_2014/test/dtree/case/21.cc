#define DTREE_CONFIG EndSeq<WithReverse<WithReverse<WithReverseBy<WithValue<dtree::DpAdd<double>, WithAggr<FreeSemigroup, NullPredicate_, WithAggr<FreeSemigroup, NullPredicate_, WithValue<FreeGroup, Begin > > > > > > > >
#include "test/dtree/seq.cc"
