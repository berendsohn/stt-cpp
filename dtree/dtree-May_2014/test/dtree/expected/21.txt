<<<
  1: v102,000,b60,000 vL4R-2F1
  10: vdb7,000,d0c,000 vL-3R5F0
  13: v6df,000,bd2,000 vL-7R25F1

  4: v137,000,475,000 vL38R-5F1
  11: v23b,000,99f,000 vL13R2F1
  7: v75e,000,410,000 vL3R-1F1
  14: v358,000,45e,000 vL15R-22F1
  6: v529,000,7c3,000 vL14R-9F0
  3: vae5,000,990,000 vL61R-1F0
  8: vc14,000,20c,000 vL5R-11F0
  2: vccc,000,fd5,000 vL4R-16F1
  15: v979,000,98b,000 vL0R1F1
  9: vd7d,000,507,000 vL6R-9F1

  5: v25c,000,698,000 vL17R11F0

  12: vcd3,000,0aa,000 vL2R-7F1
>>>
<<<
  @5: v25c,000,698,000 a000,000 a000,000 vL17R11F0 f1 f1

  @12: vcd3,000,0aa,000 a000,000 a000,000 vL2R-7F1 f0 f1

  @13: v6df,000,bd2,000 a000,000 a000,000 vL-7R25F1 f1 f1
      /1: v102,db7,b60,d0c a102,b60 a102,b60 vL7R-7F1 f0 f0
    \10: vdb7,6df,d0c,bd2 adb7,d0c adb7,d0c vL-20R4F1 f1 f1

    /9: vd7d,979,507,98b ad7d,507 ad7d,507 vL-10R6F0 f1 f1
  @15: v979,000,98b,000 a979,98b a979,98b vL0R1F1 f1 f1
      /7: v75e,358,410,45e a75e,410 a75e,410 vL21R-12F0 f1 f0
          /11: v23b,137,99f,475 a23b,99f a23b,99f vL7R-25F0 f0 f1
        \4: v137,75e,475,410 a137,475 a137,475 vL-4R35F0 f0 f0
    \14: v358,979,45e,98b a358,45e a358,45e vL-23R15F0 f1 f0
      \2: vccc,358,fd5,45e accc,fd5 accc,fd5 vL6R-11F0 f1 f0
        \8: vc14,ccc,20c,fd5 ac14,20c ac14,20c vL5R1F1 f1 f0
          \3: vae5,c14,990,20c aae5,990 aae5,990 vL56R10F0 f0 f0
            \6: v529,ae5,7c3,990 a529,7c3 a529,7c3 vL-47R-8F0 f0 f1
>>>
